Util/channel/src/main/java/io/deephaven/util/channel/BaseSeekableChannelContext.java

```diff
@@ -8,29 +8,51 @@
 import org.jetbrains.annotations.Nullable;
 
 import javax.annotation.OverridingMethodsMustInvokeSuper;
+import java.util.HashMap;
+import java.util.Map;
 import java.util.function.Supplier;
 
 public class BaseSeekableChannelContext implements SeekableChannelContext {
+
     /**
-     * An opaque resource object hosted by this context.
+     * A sentinel value to indicate that a resource is {@code null}.
      */
-    private SafeCloseable resource;
+    private static final SafeCloseable NULL_SENTINEL = () -> {
+    };
+
+    /**
+     * An empty cache of resources.
+     */
+    private static final Map<String, SafeCloseable> EMPTY_CACHE = Map.of();
+
+    /**
+     * A cache of opaque resource objects hosted by this context.
+     */
+    private Map<String, SafeCloseable> resourceCache = EMPTY_CACHE;
 
     @Override
     @Nullable
-    public final SafeCloseable apply(@NotNull final Supplier<SafeCloseable> resourceFactory) {
+    public final <T extends SafeCloseable> T getCachedResource(
+            final String key,
+            @NotNull final Supplier<T> resourceFactory) {
+        final Map<String, SafeCloseable> localResourceCache = resourceCache == EMPTY_CACHE
+                ? resourceCache = new HashMap<>(1)
+                : resourceCache;
+        SafeCloseable resource = localResourceCache.get(key);
```

Comment:
The "local" pattern is typically only used in concurrent situations where the field is `volatile`. In this situation it doesn't seem applicable? (Regardless, if the local pattern was necessary, I would expect it to also be used for the `put`)